### DISCUSSION TITLE
Create a set of mutation functions in atom

### DISCUSSION THREAD

**Discussion by moshe-kabala** [pmndrs/jotai#1270] [Ideas]
Hi, everyone new here :)
In the last few days I have been comparing some state management libraries,  I compare some use cases between them. and one of them is the ability to share mutation functions, but I don't yet find a proper way to do it in Jotai. 
so in short what I suggest is enable to create a set of mutation functions in an atom, the same as a selector in recoil, [here](https://recoiljs.org/docs/api-reference/core/selector#returning-objects-with-callbacks) 
for example

```typescript
import { atom } from "jotai";

type Item = { id: string; name: string };
let globalId = 1;

export const todoListState = atom<Item[]>([{ id: `0`, name: "somename" }]);

const logs = atom<string[]>([]);

const logger = atom((_get, set) => {
  return (msg: string) => {
    set(logs, (logs) => [...logs, msg]);
  };
});

export const todoListActions = atom((get, set: any) => {
  const log = get(logger);
  const self = {
    addByName(name: string) {
      const id = String(globalId++);
      const item = { id, name };
      return self.add(item);
    },
    add(item: Item) {
      set(todoListState, (list) => [...list, item]);
      log(`Add item (${item.name}, ${item.id})`);
      return item;
    },
    editName(id: string, name: string) {
      set(todoListState, (list) =>
        list.map((item) => (item.id === id ? { ...item, name } : item))
      );
      log(`Update item (${name}, ${id})`);
    },
    remove(...ids: string[]) {
      const setIds = new Set(ids);
      set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
      log(`Remove item/s ids: (${ids})`);
    },
    addAndRemove() {
      const { id } = self.addByName("You should not see mee");
      self.remove(id);
    },
    async addAndRemoveAsync() {
      const { id } = self.addByName("Will remove");
      await sleep(3000);
      self.remove(id);
    },
  };
});
```
The option I have now force me to split my functions like that
```typescript
export const addItem = atom<null, Item>(null, (get, set, item) => {
  set(todoListState, (list) => ({ ...list, [item.id]: item }));
  set(logger, `Add item (${item.name}, ${item.id})`);
  return item;
});

export const addItemByNam = atom<null, string>(null, (get, set, name) => {
  const id = String(globalId++);
  const item = { id, name };
  return set(addItem, item);
});
```
Or write my function as a reducer which I didn't like. 
All I miss right now is to expose the set function. I know it does not make sense because of getting function is not supposed to write, but its return functions will do, In Recoil, the returned mutation functions are wrapped. but it is a little bit annoying to wrap each function.

**Comment by dai-shi** [comment]
This is sometimes asked even in one of other libs: [react-tracked](https://github.com/dai-shi/react-tracked)
So, I know some people like such a pattern, but...

Technically, the library or React wants to control the timing of invoking update functions. Thus, `getCallback` is introduced in recoil, I guess.
Jotai's core API should be minimalistic, so we shouldn't do like recoil.

> The option I have now force me to split my functions like that

That's a proper option.

> Or write my function as a reducer which I didn't like.

It's another good one. You can create a custom hook to return mutation functions that translate to reducer actions.

Or, probably `useAtomCallback` may work in that style.

That said, we are open for ideas to add a custom hook in `jotai/utils` to cover such use cases. (I noticed `useAtomCallback` is too powerful, so a more specific hook makes sense.)

**Comment by moshe-kabala** [comment]
Hi, thanks for your answer :)

The problems I have with separating into a function are:
1. I need a custom API to call my function - I prefer a minimal interface with the library
2. Wrapping each function instead of a regular function - I prefer a minimal interface with the library
3. Multi unfortunately export and import in my code
4. Multi `useAtom` in my component
5. Because an atom with functions usually does not change it also performs better
So group functions that work on the same atom make sense to me

regarding `useAtomCallback` I can't really share the functions with it and creating a set of functions is also a challenge here

I understand the need to know when the function is invoked but it can be done by just wrapping the set and get function.

> That said, we are open for ideas to add a custom hook in jotai/utils to cover such use cases. (I noticed useAtomCallback is too powerful, so a more specific hook makes sense.)

It will be nice to see something like
`callback[s]Atom` 
```
const logs = atom<string[]>([]);

const logger = callbackAtom((_get, {set}) => {
  return (msg: string) => {
    set(logs, (logs) => [...logs, msg]);
  };
});

export const todoListActions = callbacksAtom((get, callbackInterface) => {
  const log = get(logger);
  const {set} = callbackInterface;
  return {
    add(item: Item) {
      set(todoListState, (list) => [...list, item]);
      log(`Add item (${item.name}, ${item.id})`);
      return item;
    },
    editName(id: string, name: string) {
      set(todoListState, (list) =>
        list.map((item) => (item.id === id ? { ...item, name } : item))
      );
      log(`Update item (${name}, ${id})`);
    },
    remove(...ids: string[]) {
      const setIds = new Set(ids);
      set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
      log(`Remove item/s ids: (${ids})`);
    },
  };
});
```

`callback[s]AtomFamily` 
```
const itemActions = callbacksAtomFamily(
  (id: string) => (get, callbackInterface) => {
    const { editName, remove } = get(todoListActions);
    const { set } = callbackInterface;

    return {
      editName(name: string) {
        editName(id, name);
      },
      remove() {
        remove(id);
      },
      customEdit() {
        set(todoListState, (list) =>
          list.map((item) => (item.id === id ? { ...item, name: "X" } : item))
        );
      },
    };
  }
);
```

**Reply by dai-shi** [reply]
What you suggest literally is not possible by nature and by current design, but let's look for a possible (workaround) solution.

Here's the `useAtomCallback` version:
```js
const logs = atom<string[]>([]);

const useLogger = () => useAtomCallback(useCallback((get, set, msg) => {
  set(logs, (logs) => [...logs, msg]);
}, []));

export const todoListActions = () => {
  const log = useLogger();
  return {
    add: useAtomCallback(useCallback((get, set, item) => {
      set(todoListState, (list) => [...list, item]);
      log(`Add item (${item.name}, ${item.id})`);
      return item;
    }, [log])),
    editName: useAtomCallback((get, set, { id, name }) => { // unfortunately, only one arg is supported now (but, this can be solved in some ways)
      set(todoListState, (list) =>
        list.map((item) => (item.id === id ? { ...item, name } : item))
      );
      log(`Update item (${name}, ${id})`);
    }, [log])),
    remove: useAtomCallback((get, set, ids) =>) { // again, the same
      const setIds = new Set(ids);
      set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
      log(`Remove item/s ids: (${ids})`);
    }, [log])),
  };
});
```
So, `useCallback` is a bit annoying, but how does it look otherwise? We could build `useAtomCallbacks` to make it just one hook (with `useMemo` in this case).

**Reply by moshe-kabala** [reply]
Hi, Thanks for your answer I understand that multi-mutable-functions atom can't at this moment, so your idea is probably the best possible option.

> So, useCallback is a bit annoying, but how does it look otherwise? We could build useAtomCallbacks to make it just one hook (with useMemo in this case).

If something like that is possible by wrapping the set and get function it will be a significant improvement from the developer experience 

```
const useLogger = () => useAtomCallback(useCallback((get, set, msg) => {
    set(logs, (logs) => [...logs, msg]);
  }, []));
  
export const todoListActions = () => {
    const log = useLogger();
    return useAtomCallbacks((get, set) => {
        add(item)  {
            set(todoListState, (list) => [...list, item]);
            log(`Add item (${item.name}, ${item.id})`);
            return item;
        },
        editName( id, name ) {
        set(todoListState, (list) =>
            list.map((item) => (item.id === id ? { ...item, name } : item))
        );
        log(`Update item (${name}, ${id})`);
        },
        remove(ids) { 
            const setIds = new Set(ids);
            set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
            log(`Remove item/s ids: (${ids})`);
        }
    }, [log]);
};
```

**Reply by dai-shi** [reply]
In that case, I think translating to reducer-like actions would be easier. Let's name it `useAtomActions`.

```js
const useAtomActions = (createCallbacks) => {
  const anAtom = useMemo(() => atom(null, (get, set, { type, args }) => {
    createCallbacks(get, set)[type](...args)
  }), [createCallbacks])
  const dispatch = useSetAtom(anAtom)
  return useMemo(() => Object.fromEntries(Object.keys(createCallbacks).map((type) => [type, (...args) => dispatch({ type, args })])), [createCallbacks, dispatch])
}

// usage
export const useTodoListActions = () => {
  const log = useLogger()
  return useAtomActions(useMemo((get, set) => {
    add(item)  {
      set(todoListState, (list) => [...list, item])
      log(`Add item (${item.name}, ${item.id})`)
    },
    editName( id, name ) {
      set(todoListState, (list) =>
        list.map((item) => (item.id === id ? { ...item, name } : item))
      );
      log(`Update item (${name}, ${id})`);
    },
    remove(ids) { 
      const setIds = new Set(ids);
      set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
      log(`Remove item/s ids: (${ids})`);
    }
  }, [log]));
};
```

Note that not everything you requested is implemented.

**Reply by moshe-kabala** [reply]
Assume looks good.
All I miss here is the return value by an action.

> Note that not everything you requested is implemented.

Yes it still is not a shared atom

**Reply by dai-shi** [reply]
> Assume looks good. All I miss here is the return value by an action.

This is a little tricky, but as discussed in #1245 and #1271, we avoid it for the future. Otherwise, you can just return a value.

> Yes it still is not a shared atom

As you see, all my proposition is hooks.
It's impossible to make it an atom in the current design.

**Comment by AjaxSolutions** [comment]
Can you use a hook which returns store actions?
```
const useStoreActions = () => {
    return {
        actionOne: useSetAtom(oneAtom),
        actionTwo: useSetAtom(twoAtom),
        actionThree: useSetAtom(threeAtom),
    };
};
```

**Reply by moshe-kabala** [reply]
Hi thanks, Nice idea
I see two problems here
1. The actions are not in the shared atom (using many hooks for each instance)
2. A lot of unnecessary work to wrap every function in `useAtom`  and then create a custom `useStoreActions`

**Comment by AjaxSolutions** [comment]
Why do you think #1 is a problem?  Is it going to create a performance issue?

I''m also looking for a way to organize mutation actions.  Wrapping atoms in an atom looks complex.

**Reply by moshe-kabala** [reply]
I can agree that in most cases 1 is not an issue just unnecessary useAtom.

**Comment by bjoernsteinborn** [comment]
Is this still supposed to work with the latest Version of Jotai? `createCallbacks` is once called like a function and another time iterated over as it would be an object with `Object.keys`? Or is this covered by `useReducerAtom` meanwhile anyways? Can’t see an example with payload for an action though. 

> In that case, I think translating to reducer-like actions would be easier. Let's name it `useAtomActions`.
> 
> ```js
> const useAtomActions = (createCallbacks) => {
>   const anAtom = useMemo(() => atom(null, (get, set, { type, args }) => {
>     createCallbacks(get, set)[type](...args)
>   }), [createCallbacks])
>   const dispatch = useSetAtom(anAtom)
>   return useMemo(() => Object.fromEntries(Object.keys(createCallbacks).map((type) => [type, (...args) => dispatch({ type, args })])), [createCallbacks, dispatch])
> }
> 
> // usage
> export const useTodoListActions = () => {
>   const log = useLogger()
>   return useAtomActions(useMemo((get, set) => {
>     add(item)  {
>       set(todoListState, (list) => [...list, item])
>       log(`Add item (${item.name}, ${item.id})`)
>     },
>     editName( id, name ) {
>       set(todoListState, (list) =>
>         list.map((item) => (item.id === id ? { ...item, name } : item))
>       );
>       log(`Update item (${name}, ${id})`);
>     },
>     remove(ids) { 
>       const setIds = new Set(ids);
>       set(todoListState, (list) => list.filter((id) => !setIds.has(id)));
>       log(`Remove item/s ids: (${ids})`);
>     }
>   }, [log]));
> };
> ```
> 
> Note that not everything you requested is implemented.

**Reply by dai-shi** [reply]
> another time iterated over as it would be an object with Object.keys?

I think I meant `createCallbacks()` for the second one.

It should still work with the latest Jotai, but it's not very idiomatic. It's full of hacks, and not generally recommended. But, please feel free to try it out.
